File: algorithm/#2/addTwoNumbers.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        return add(l1: l1, l2: l2, val: 0)
    }

    func add(l1: ListNode?, l2: ListNode?, val: Int) -> ListNode? {
        let (sum, next1, next2): (Int, ListNode?, ListNode?)
        switch (l1, l2) {
            case let (n1?, n2?):
                (sum, next1, next2) = (n1.val + n2.val + val, n1.next, n2.next)
            case (let n?, nil):
                (sum, next1, next2) = (n.val + val, n.next, nil)
            case (nil, let n?):
                (sum, next1, next2) = (n.val + val, n.next, nil)
            default:
                if (val > 0) {
                    return ListNode(val)
                } else {
                    return nil
                } 
        }
        return ListNode(sum % 10, add(l1: next1, l2: next2, val: sum / 10))
    }
}

File: algorithm/#2/addTwoNumbers.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode? AddTwoNumbers(ListNode? l1, ListNode? l2) {
        return Add(l1, l2, 0);
    }

    static ListNode? Add(ListNode? l1, ListNode? l2, int val)
    {
        (int sum, ListNode? next1, ListNode? next2) = (0, null, null);
        switch ((l1, l2)) 
        {
            case (ListNode n1, ListNode n2):
                (sum, next1, next2) = (n1.val + n2.val + val, n1.next, n2.next);
                break;
            case (ListNode n, null):
                (sum, next1, next2) = (n.val + val, n.next, null);
                break;
            case (null, ListNode n):
                (sum, next1, next2) = (n.val + val, n.next, null);
                break;
            default:
                if (val > 0)
                {
                    return new ListNode(val);
                }
                else
                {
                    return null;
                }
        }
        var node = new ListNode(0);
        node.val = sum % 10;
        node.next = Add(next1, next2, sum / 10);
        return node;
    }
}